### DISCUSSION TITLE
Identity Foundation Error

### DISCUSSION THREAD

**Discussion by jwanyeki** [Azure/azure-saas#217] [Q&A] [answered]
I am running the ./run.sh command for the identity foundation but the script is completing with the following error:

Identity Foundation deployed with errors. Sometimes this is due to an error happening with the Azure Resource Manager. Try running the script once more to see if it continues past the point of failure.

I have run the script severally but am still getting the same error. Any help will be highly appreciated.

**Comment by 1iveowl** [comment]
Thanks @jwanyeki for reaching out. 

Can you share some more details? What does the log say? Is there a specific part that fails?

**Reply by jwanyeki** [reply]
[deployment_error.txt](https://github.com/Azure/azure-saas/files/11401635/deployment_error.txt)
![asdk_error](https://user-images.githubusercontent.com/38181921/236331908-b116ad94-a37a-4230-bf54-0e787950d9a3.png)

**Reply by jwanyeki** [reply]
Thanks. See some of the logs and deployment progress. All the previous steps executed successfully.

**Reply by jwanyeki** [reply]
Am online? I can share my client

**Reply by 1iveowl** [reply]
Thanks for sharing. That’s helpful. Looks like the script is not pulling the certificate-policy.json for some reason. I need to do some more investigation on this. I will get back to you as soon as I understand more about why this might be happening, hopefully by tomorrow.

**Reply by jwanyeki** [reply]
Thank you Sir for the quick response. Looking forward for more insights from your analysis. I appreciate your support.

**Reply by 1iveowl** [reply]
Hi @jwanyeki,

I see that the script asks you to to log in. Did you not run this command before running the script?
```bash 
az login --scope "https://graph.microsoft.com/.default
``` 
Please see the [guide](https://github.com/Azure/azure-saas/tree/main/src/Saas.Identity/Saas.IdentityProvider#logging-into-az-cli). 

I'm unable to recreate the error based on the above, so if this doesn't help, could I convince you to try and start over again - i.e., delete all your local files and repo and clone from your forked version of ASDK again. Then follow the [Quick Start](https://azure.github.io/azure-saas/quick-start/)?

Also, please make sure that you're running the latest version of Docker Desktop and AZ CLI on your development machine.

The issue you are facing is that the file `certificate-policy.json`, which is placed in the `./config` directory, is not downloaded. This file should be downloaded automatically by the script. When I run this script on a clean machine this works as expected.

**Reply by jwanyeki** [reply]
Thanks alot for the advise. I have restarted the process but the error has persisted. Could it be an issue with WSL. I am running the scripts from ubuntu terminal on windows 10 with the latest Docker Desktop 4.19.0  and azure-cli version 2.48.1.

**Reply by 1iveowl** [reply]
Which version of Ubuntu are you using?

It's been testet with Ubuntu 22.04 running w/ WSL

**Reply by jwanyeki** [reply]
I am using 22.04.2 LTS.

**Reply by 1iveowl** [reply]
Hmmm. Strange. That should work fine. 

Are you still asked to login when you run the script after running the Az command I mention above?

**Reply by jwanyeki** [reply]
I have tried the process 10 times today without success. The error message on the logs is still the same
cp: cannot stat '/asdk/src/Saas.Identity/Saas.IdentityProvider/deployment/config/certificate-policy.json': No such file or directory

**Reply by jwanyeki** [reply]
> Hmmm. Strange. That should work fine.
> 
> Are you still asked to login when you run the script after running the Az command I mention above?

Yes. Am prompted for the login and the code

**Reply by 1iveowl** [reply]
You should not be asked to login again. Could you try and uninstall az cli and install it again? It feels like a login issue. 

I'm really sorry that you are having this issue. I need to look closer at the script again and maybe send you something to test.

Anyhow, it's 7:30pm here in my timezone and I'm not by my computer currently. I will get back to you.

**Reply by jwanyeki** [reply]
Thanks. let me try the workaround. My time is 2030hrs, Nairobi. I will keep trying for the next 5 hours just in case you have some more input. I really want to adopt this SDK in my SaaS Application.

**Reply by jwanyeki** [reply]
> You should not be asked to login again. Could you try and uninstall az cli and install it again? It feels like a login issue.
> 
> I'm really sorry that you are having this issue. I need to look closer at the script again and maybe send you something to test.
> 
> Anyhow, it's 7:30pm here in my timezone and I'm not by my computer currently. I will get back to you.

I have reinstalled az cli and reinitialized the local files from the forked repository but the certificate-policy.json exception is still being generated.

**Comment by jwanyeki** [comment]
![run_error](https://user-images.githubusercontent.com/38181921/236548383-3a969f9d-1bbd-4fea-bf70-231e5cd66c9d.png)

**Reply by jwanyeki** [reply]
Can the certificate-policy.json be manually generated and pasted to the deployment folder? Or is there an alternative way of deploying the sdk resources?

**Reply by 1iveowl** [reply]
Yes. Here's the content of the file. Just copy this into file wth the name `certificate-policy.json` and place the file in the `./config` directory,

I doubt it will fix the underlying issue, but it's worth a try and it might help us figure out what's causing this. 

```json
{
  "issuerParameters": {
    "certificateTransparency": null,
    "name": "Self"
  },
  "keyProperties": {
    "curve": null,
    "exportable": true,
    "keySize": 4096,
    "keyType": "RSA",
    "reuseKey": true
  },
  "lifetimeActions": [
    {
      "action": {
        "actionType": "AutoRenew"
      },
      "trigger": {
        "daysBeforeExpiry": 90
      }
    }
  ],
  "secretProperties": {
    "contentType": "application/x-pkcs12"
  },
  "x509CertificateProperties": {
    "keyUsage": [
      "cRLSign",
      "dataEncipherment",
      "digitalSignature",
      "keyEncipherment",
      "keyAgreement",
      "keyCertSign"
    ],
    "subject": "CN=asdktest99yq.onmicrosoft.com",
    "validityInMonths": 12
  }
}
```

**Reply by jwanyeki** [reply]
Thank you very much for your prompt response. I have created the file but am now getting initialization error as shown on the screenshot here.
![initialization_error](https://user-images.githubusercontent.com/38181921/236554753-37c79734-d0a8-49ed-9a61-f684bc7f8da0.png)

**Reply by 1iveowl** [reply]
Could you try this:
1.  Edit the last line of the script `start.sh` by adding a `#` right after `bash`. The last like of the script should look like this after the edit:
```bash
bash # /asdk/src/Saas.Identity/Saas.IdentityProvider/deployment/start.sh
```
2. Now run the script `./run.sh`. This will take you to a command prompt inside the container.
3. From this prompt run the az cli command `az account show` and tell me what you see?

What you should see is something like this: 
```json
{
  "environmentName": "AzureCloud",
  "homeTenantId": "<some guid>",
  "id": "<some guid>",
  "isDefault": true,
  "managedByTenants": [],
  "name": "<Tenant Name, string>",
  "state": "Enabled",
  "tenantId": "<some guid>",
  "user": {
    "name": "<your name>",
    "type": "user"
  }
}
```

**Reply by jwanyeki** [reply]
Hello,

Thank you very much for your continued support. The bash /asdk/src/Saas.Identity/Saas.IdentityProvider/deployment/start.sh script line is at the tail of my run.sh and not start.sh. I have added a # right after bash, executed az login --tenant <my tenant id> and then az account show and  the tenant information was displayed as expected. See the  script below where security sensitive details have been replaced with XX

{
    "cloudName": "AzureCloud",
    "homeTenantId": "XXX",
    "id": "XXX",
    "isDefault": true,
    "managedByTenants": [],
    "name": "XXX",
    "state": "Enabled",
    "tenantId": "XXX",
    "user": {
      "name": "XXX",
      "type": "user"
    }
  }
Am still looking for a solution to the technical hitch

**Reply by 1iveowl** [reply]
You are correct. I meant `run.sh`

Please don’t run `az login —tenant`first. The point is that this should not be necessary, since you already ran `az login`outside the container. The container should be able to “pull in” your credentials. If it’s not, then that is what is causing the the issue.

Could you please exit the container run it again and only run `az account show`and let me know what you see?

**Reply by 1iveowl** [reply]
Btw. This is how the container get’s your existing az cli credentials:  `--volume "${HOME}/.azure/":/asdk/.azure:ro \`.

You may want to check and see if you can access the directory `./asdk/.azure/` from inside the container. It should hold a number of the az cli config files.

**Reply by jwanyeki** [reply]
Hello @1iveowl . I concur. This seems to be an az-cli issue. I have executed the requested command but an being prompted to re-authenticate. What do you suggest I do. I had followed the cli-installation guidelines provided [here](https://learn.microsoft.com/en-us/cli/azure/install-azure-cli-linux?pivots=apt).

**Reply by jwanyeki** [reply]
> Btw. This is how the container get’s your existing az cli credentials: `--volume "${HOME}/.azure/":/asdk/.azure:ro \`.
> 
> You may want to check and see if you can access the directory `./asdk/.azure/` from inside the container. It should hold a number of the az cli config files.

Yes. The directory files are accessible on the running container.

**Comment by jwanyeki** [comment]
![az-error](https://user-images.githubusercontent.com/38181921/236629229-a5e14782-a30b-459f-bd11-459b6437d629.png)

**Reply by 1iveowl** [reply]
Quick question. Are you running WSL 2?

You could also try and reset WSL 2 starting from an all new version of Ubuntu 22.04?

**Reply by jwanyeki** [reply]
Hello @1iveowl? Yes. I tried resetting WSL on windows 10 but the error is still persistent. I have even tried repeating the entire installation process for foundation services on a different machine with windows 11 and the same errors were generated. Could there be something missing on the documentation?

**Reply by 1iveowl** [reply]
We've tested this on multiple machines without running into this issue. So I'm still struggling to figure out what is at play here. 

What is your role and your permissions on the Azure subscription and tenant you are logging into? Are you the "Service Administrator" on the subscription? You can see this by going to you Subscription | Access Control (IAM) and pressing the "View my" access button.
![image](https://user-images.githubusercontent.com/12871120/236703634-140c607b-85bd-486d-a711-bd7a23e3cb7a.png)

**Reply by jwanyeki** [reply]
Hello @1iveowl . I am the service administrator. See the screenshot attached
![service admin](https://user-images.githubusercontent.com/38181921/236703853-dac52ff6-c29e-452a-828e-26b5ab6c92df.png)
.

**Reply by jwanyeki** [reply]
I have finally managed to deploy the identity foundation services. There seems to be some missing nitty gritties on the available documentation. I will share my experience just in case someone else is stuck. Thanks @1iveowl  for the incredible support.

**Reply by 1iveowl** [reply]
@jwanyeki I'm so happy to hear this! I've been pulling my hair all weekend trying to figure out what might have been causing this. 

Please let me know what the issue was and I will make sure to implement a fix or fix the documentation or both, so that others will not have to go through your experience. 

And thank you so much for your patience, in this matter!

**Reply by jwanyeki** [reply]
Thanks @1iveowl . The issue was on the Azure-CLI configuration, and location settings on the config.json. I will share a small document tomorrow on the clarity needed when configuring CLI on WSL as well as location settings and why the respective documentation sections need improvement for a smoother deployment. I also faced a few challenges with GH_Token. The scripts work well though and many thanks to the team that build this.

**Reply by 1iveowl** [reply]
> Thanks @1iveowl . The issue was on the Azure-CLI configuration, and location settings on the config.json. I will share a small document tomorrow on the clarity needed when configuring CLI on WSL as well as location settings and why the respective documentation sections need improvement for a smoother deployment. I also faced a few challenges with GH_Token. The scripts work well though and many thanks to the team that build this.

Awesome! Can't wait to learn more about what you learned so that we can improve the documentation. Great work that you figured this out!

**Reply by 1iveowl** [reply]
@jwanyeki

 Hi James. If you don't have time to write a detailed document, feel free to dump the essential finding here and I'll be happy to write up the documentation changes needed.

**Reply by jwanyeki** [reply, marked as answer]
Hello @1iveowl . Thanks for the follow-up. Below are my power points  for a smooth deployment which might not be clear from the documentation
1. The location specified on the initConfig section of the config.json file must be a valid name within the locations displayed once you run the command (az account list-locations --output table). This must be within the **name column** of the output and **not the displayname** column e.g francecentral is valid but using France Central will prevent the certificate-policy.json file from being generated.
2. Uninstall previous versions of CLI and follow the installation guide provided [here](https://learn.microsoft.com/en-us/cli/azure/install-azure-cli). If you had installed the windows msi version of the az-cli previously while working with PowerShell, your container might be unable to pull az-cli credentials if proper configurations are not done within WSL. In my case I uninstalled all previous versions of az-cli and run(
curl -sL https://aka.ms/InstallAzureCLIDeb | sudo bash)  for the container to persist the credentials.
3. Uninstall all previous versions of gh cli before running the gh cli **install** command provided [here](https://github.com/cli/cli/blob/trunk/docs/install_linux.md) . I was running on version 2.4 of gh cli that caused the deployment to generate GH_Token exceptions and running  **gh auth token** command could not generate the authentication token required to complete the deployment. Remember to run gh --version to confirm your gh cli version before attempting to execute ./run.sh. The current stable version of gh cli which is yielding a seamless deployment is [2.29.0](https://github.com/cli/cli/releases/tag/v2.29.0)

**Reply by 1iveowl** [reply]
This is brillant @jwanyeki! Thanks for the contribution. I will go and update the documentation with this right away!

**Reply by 1iveowl** [reply]
The documentation enhancements have been implemented.
